### ISSUE TITLE
Speedup pairwise inference for validation using DDP and handling duplicates 

### ISSUE THREAD

**Issue by AlekseySh** [OML-Team/open-metric-learning#267]
blocked by: https://github.com/OML-Team/open-metric-learning/issues/267 (because we need the ids of the samples to remove duplicates in DDP regime)

**Comment by AlekseySh** [comment]
Blocked by: https://github.com/OML-Team/open-metric-learning/pull/266

**Comment by AlekseySh** [comment]
draft 

```
ef ids_of_first_entrance(x):
    seen = set()
    first_ids = []
    for i, el in enumerate(x):
        if el not in seen:
            first_ids.append(i)
        seen.add(el)
    return first_ids


def keep_only_records_with_unique_ids(ids, x, sort_by_ids=True):
    ids_first = ids_of_first_entrance(ids)
    ids = [ids[i] for i in ids_first]
    x = x[ids_first]

    if sort_by_ids:
        ii_permute = torch.argsort(torch.tensor(ids))
        x = x[ii_permute]
        ids = [ids[i] for i in ii_permute]

    return ids, x


# todo: use lightning here to work with ddp and half precision
# fmt: off
def pairwise_inference(
        model: IPairwiseModel,
        dataset: IPairsDataset,
        num_workers: int,
        batch_size: int,
        verbose: bool
) -> Tensor:
    model.cuda()
    prev_mode = model.training
    model.eval()
    loader = DataLoader(dataset, batch_size=batch_size, num_workers=num_workers, shuffle=False)
    print(len(loader), "LEN")
    loader = patch_dataloader_to_ddp(loader)
    device = get_device(model)

    loader = tqdm(loader) if verbose else loader

    outputs = []
    ids = []
    with torch.no_grad():
        for batch in tqdm(loader):
            x1 = batch[dataset.pair_1st_key].to(device)
            x2 = batch[dataset.pair_2nd_key].to(device)
            outp = model(x1=x1, x2=x2)
            outputs.append(outp)
            ids.extend(batch['idx'].tolist())

    model.train(prev_mode)
    # return [torch.cat](http://torch.cat/)(outputs).detach().cpu()
    preds = [torch.cat](http://torch.cat/)(outputs).detach().cpu()
    res = {"preds": preds, "ids": ids}
    res = sync_dicts_ddp(res, 2)
    preds, ids = res['preds'], res['ids']

    ids, preds = keep_only_records_with_unique_ids(ids=ids, x=preds)

    print(len(preds), ' AFTER')
    return preds
    ```
